### PR TITLE
cleanup: collect getResourceRequest logic of scheduler

### LINKED ISSUE #90402: Make the "GetResourceRequest" logic in scheduler pkg more clean
**> This issue is for tracking some problems I found and I plan to fix.**

**Why is this needed**:
During I fixed the bug [**bugfix: initcontainer is also taken into account when calculating resource requests**](https://github.com/kubernetes/kubernetes/pull/89222), I found some issues of calculating resource request logic existed in scheduler package.  

1. **Duplication**: Actually, in Kubernetes code base, only kube-scheduler gives the default resource request to every container for calculating score purpose. So we have our own logic. But there is a [unified function](https://github.com/kubernetes/kubernetes/blob/5713c22eecff4610026643fbd3d37c33a43c168d/pkg/api/v1/resource/helpers.go#L133) which also can be used to get the resource request of pod  in `/pkg/api/v1/resource` package. 

2. **Default resource request value**: In the [current implementation](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/noderesources/resource_allocation.go#L118), we check every container of one pod if it specifies the explicit resource request. If there is no explicit value, we will give a default resource request value to the **container**. Actually, we can change the default value of per container to per pod. It likes 
```go
resourceRequest := max(sum(pod.Spec.Containers), max(pod.Spec.InitContainers))
if  resourceRequest== 0 {
              resourceRequest = defaultResourceRequestValue
}
```
3. **Explicit zero resource request**:  In this [unit test](https://github.com/kubernetes/kubernetes/blob/3538320d74f98fc0972206d20cd3d76186d88aa9/pkg/scheduler/framework/plugins/noderesources/least_allocated_test.go#L100), it specifies the explicit zero resource request. It looks like this test doesn't want to get the default resource request. It should be considered carefully ****If we should give the default resource request value** for the pod which specifies the explicit zero value**. [Context](https://github.com/kubernetes/kubernetes/pull/89222#discussion_r411617331)

4. **IsScalarResourceName**: In our [current of calculating non zero resource request logic](https://github.com/kubernetes/kubernetes/blob/b850b5ce2e7d282abae01a26be4d14d56d71bfad/pkg/scheduler/util/non_zero.go#L76), we consider if the resource belongs to the **scalar resource**. Not sure if it is valuable to remain. Because the [GetResourceRequest](https://github.com/kubernetes/kubernetes/blob/5713c22eecff4610026643fbd3d37c33a43c168d/pkg/api/v1/resource/helpers.go#L133) of `/pkg/api/v1/resource` package doesn't consider it.

5. **Lack of correct test cases**: When I fixing the bug I mentioned, I found it is hard to add the test logic about initContainer in [TestNodeInfoRemovePod](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/v1alpha1/types_test.go#L814). Because it uses the [MakeBasePod](https://github.com/kubernetes/kubernetes/blob/3e2ae631b4caef5f823a5e121f860c653da32ff5/pkg/scheduler/framework/v1alpha1/types_test.go#L247) to construct the pod test cases. The current logic of  [MakeBasePod](https://github.com/kubernetes/kubernetes/blob/3e2ae631b4caef5f823a5e121f860c653da32ff5/pkg/scheduler/framework/v1alpha1/types_test.go#L247) is not good for me. We should refactor it and make the test for "GetResourceRequest" logic more complete.

In my [bugfix PR](https://github.com/kubernetes/kubernetes/pull/89222), for easier to cherry-pick to the older version, I just focus on fixing the bug based on the current implementation.  So I think we need to do the improvement in following PRs.

### PR DESCRIPTION
**Pull request by pravarag**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
As part of this change, we are trying to collect all the `GetResourceRequest` logic of scheduler in one place.
#### Which issue(s) this PR fixes:

Fixes #90402

#### Does this PR introduce a user-facing change?
NONE

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/component-helpers/scheduling/corev1/nodeaffinity"
+	schedutil "k8s.io/kubernetes/pkg/scheduler/util"
```
Comment (alculquicondor) [inline staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go:24]: component-helpers can't depend on kubernetes/pkg

File: staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go [hunk 2]
```diff
@@ -99,3 +100,47 @@ func getFilteredTaints(taints []v1.Taint, inclusionFilter taintsFilterFunc) []v1
...
 }
+
+// calculatePodResourceRequest returns the total non-zero requests. If Overhead is defined for the pod and the
+// PodOverhead feature is enabled, the Overhead is added to the result.
+// podResourceRequest = max(sum(podSpec.Containers), podSpec.InitContainers) + overHead
+func calculatePodResourceRequest(pod *v1.Pod, resource v1.ResourceName) int64 {
+	var podRequest int64
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		value := schedutil.GetNonzeroRequestForResource(resource, &container.Resources.Requests)
```
Comment (Huang-Wei) [inline staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go:111]: You can move the dependent logic in this file.

**Comment by dims** [comment]
please reopen if you still need this WIP PR. Trying to declutter the open PR(s)

**Comment by alculquicondor** [comment]
More thoughts on this: #106769

**Comment by pravarag** [comment]
My apologies for not being able to spend enough time on working this out, I'll reopen it and finish it.

**Comment by alculquicondor** [comment]
@yuzhiquan is already working on it

**Pull request closed without merging**